[lldb] Move Transport class into lldb_private (NFC)

JDevlieghere opened the pull request:
Move lldb-dap's Transport class into lldb_private so the code can be shared between the "JSON with header" protocol used by DAP and the JSON RPC protocol used by MCP (see [1]).

[1]: https://discourse.llvm.org/t/rfc-adding-mcp-support-to-lldb/86798

JDevlieghere committed:
[lldb] Move Transport class into lldb_private (NFC)

Move lldb-dap's Transport class into lldb_private so the code can be
shared between the "JSON with header" protocol used by DAP and the JSON
RPC protocol used by MCP (see [1]).

[1]: https://discourse.llvm.org/t/rfc-adding-mcp-support-to-lldb/86798

JDevlieghere commented:
@ashgti Let me know if you think it's too early to land this change. Given the popularity of JSON RPC, I think this would be generally useful, but let me know if you'd like me to hold off until the RFC has converged.

ashgti commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -0,0 +1,131 @@
+/// A transport class that uses JSON for communication.
+class JSONTransport {
+public:
+  JSONTransport(llvm::StringRef client_name, lldb::IOObjectSP input,
```
Maybe we should hoist client_name out of the transport class itself?

Our logging helper could wrap the client name into the logger instance instead of associating it directly to the transport or we could move the client name down to the `lldb_dap::Transport`. That might be worth thinking about but not changes are required.

ashgti commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -0,0 +1,131 @@
+  virtual llvm::Expected<std::string>
+  ReadImpl(const std::chrono::microseconds &timeout) override;
+
+  static constexpr llvm::StringLiteral kHeaderContentLength =
+      "Content-Length: ";
+  static constexpr llvm::StringLiteral kHeaderSeparator = "\r\n\r\n";
```
I've been meaning to revisit this, but the DAP spec doesn't actually require the content length to be formatted precisely like `Content-Length: \d+\r\n\r\n` its actually more like an HTTP header. I've seen other DAP implementations include other headers like 
`Content-Length: 123\r\nContent-Type: application/vscode-jsonrpc; charset=utf-8\r\n\r\n` (the dart DAP implementation does this). Its assumed that the header is a set of `<key>: <value>\r\n` and is delimited by an empty line `\r\n` like HTTP headers are done. The only meaningful key is 'Content-Length' though. See https://microsoft.github.io/debug-adapter-protocol/overview 'Header Part'

This is a long winded way of saying, maybe we should call this `HTTPDelimitedJSONTransport`

ashgti commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -0,0 +1,131 @@
+    if (!message)
+      return message.takeError();
+    return llvm::json::parse<T>(/*JSON=*/*message,
+                                /*RootName=*/"transport_message");
```
Should the `RootName` label come from the class somehow since I think this would show up in the error messages if a malformed value was encountered.

ashgti reviewed (approved):
Looks good to me!

JDevlieghere committed:
Address John's feedback

 - Move client name out of the base class
 - Use HTTPDelimitedJSONTransport
 - Don't set the root name
